### PR TITLE
feat(datasource/graphene)

### PR DESCRIPTION
**Pull request by chrisj**
continued from https://github.com/google/neuroglancer/pull/317

refactored to use SegmentationGraphSource. still very much in progress

### CODE REVIEW

**Comment by chrisj** [comment]
I have some questions for how I should proceed.

First to summarize what I did and open to feedback. I added a createRenderLayer method to the graph connection interface. SegmentationUserLayer calls it and adds a layer if one exists. Maybe it should be an array to be flexible. I added the ability to override `select` if there is a graph and it implements a select operation.

Many of our operations need to be able to change the VisibleSegmentState. Do you suggest we use the connections set pattern that you used in local? I started to do that but I wasn't sure the case where we would have more than one graph connection. This is also related to the next question.

The `select` method is currently it is on the graphConnection but I think it should be moved it to graph because it shouldn't be different from merge and split. Though I'm also a little split because this is called from the SegmentationUserLayer which has a single graphConnection associated with it. Regardless of using the connections set, it still relies on accessing our renderLayer. I haven't implemented the functionality in this seunglab PR https://github.com/seung-lab/neuroglancer/commit/5045538e84b2388c2c06078fc3467d2fd4a7167a#

But basically we want to prevent the select operation if the user is zoomed out too far because it would put too much stress on the browser and our graph backend. It checks the zoom by comparing the sliceview pixel size with the layer voxel size. I'm still trying to understand this code, I think it should be possible to simplify it and I was hoping we could do this calculation inside of our select method in our graph implementation. I was struggling to figure out how to access the current pixel size given a SegmentationUserLayer. Maybe you have a better idea on how to deal with this issue.

We have custom UI for graphene and I'm not sure how to integrate custom configurable UI limited to the Graphene source. For example, we have a simple date picker in our graph tab that allows us to view the graph as it was at a previous point in time (we also disable editing if that is set). We probably want to add a tab for all of our graphene specific UI. Is this already possible in neuroglancer? I haven't spent much time in this area.

**Comment by jbms** [comment]
The idea of the SegmentationGraphSourceConnection is precisely to allow the graph code to manipulate the selected segment set and the equivalences map.

Note that if you add the same graphene source to multiple SegmentationUserLayers, the same SegmentationGraphSource object will be shared by multiple layers.

Regarding `select`: what I did in the prototype nggraph source is that operations that select and deselect segments still just modify the visibleSegments set directly, but the SegmentationGraphSourceConnection sets up signal handlers that monitor changes in the visibleSegments set, and if it finds a base supervoxel id in the visibleSegments set, queries the server to determine the corresponding root id and then replaces it in the visibleSegments set (and also updates the equivalences map as needed).

Regarding the chunked graph source and zooming out too far, my understanding is the intention is to avoid loading equivalence data when zoomed out too far, right?  Not specifically related to select?  In filterVisibleSources you could exclude sources when zoomed out too far.

As far as adding custom UI controls, currently there is no interface for a graph connection to do that.  One could be added, either for adding existing controls to one of the tabs, or as a separate tab.  But where it makes sense, I think it would be better to instead the feature as an optional function exposed by the GraphConnection (which would initially only be supported by graphene), and only add the control when the feature is available (i.e. when using a graphene source).  For example, I think the time setting could make sense to expose as an optional GraphConnection feature.  And the graphene-specific split and merge actions could also be exposed as additional split and merge tools.

File: src/neuroglancer/viewer.ts [hunk 1]
```diff
@@ -696,11 +696,61 @@ export class Viewer extends RefCounted implements ViewerState {
...
+   * Called after an edit is made to "deactivate" merge/split
+   */
+   deactivateEditMode() {
+    if (this.mouseState.actionState === ActionState.INACTIVE) {
+      this.mouseState.toggleAction();
+    }
+    this.handleModeChange(this.mouseState.actionMode === ActionMode.MERGE);
+  }
+
+  handleModeChange(merge: boolean) {
```
[the remark this answers is not in the document]
Comment (chrisj) [reply src/neuroglancer/viewer.ts:709]: I left that code in there because I wasn't sure the direction to go for to handle graphene specific UI. I haven't touched it since adapting to your recent changes.

Also I I forgot to mention merge and split. We send the xyz coordinates to the backend for the merge and split points. We also have our multicut feature were we can select multiple points for a and b so our server can compute a better split point. So you would suggest we implement this as additional tools rather than change/add to the existing split+merge methods?

[the remark this answers is not in the document]
Comment (chrisj) [reply src/neuroglancer/viewer.ts:709]: didn't see your above comment, reading that now...

[the remark this answers is not in the document]
Comment (jbms) [reply src/neuroglancer/viewer.ts:709]: It seems like merge could potentially be modified if the only change is the addition of the position --- it seems like it might be reasonable to always provide the position, and graph sources that don't care can just ignore it.  For split it would probably make more sense to create a separate tool since there are a lot of differences.  Then the SegmentationGraphSource could indicate which types of splits it supports.

To provide the position, you will need a RenderLayerTransform associated with the SegmentationGraphSourceConnection  so that the position can be transformed to the appropriate coordinate space --- you could add a `transform: WatchableValueInterface<RenderLayerTransformOrError>` parameter to `SegmentationGraphSource.connect`.

**Comment by chrisj** [comment]
filterVisibleSources to limit zoom out looks like a good solution, thanks, and I think I understand the graph source vs graph connection now.
I'm working on the segment query.

What was your idea behind watchSegment, I'm wondering if that could be useful for graphene.

**Comment by jbms** [comment]
trackSegment is actually currently unused, so you can ignore that.

I think I originally added it for use by the split and merge tools to keep track of the root id corresponding to the selected seed points (which might change as other splits/merges are performed), but then found it was not necessary.

In the nggraph prototype all segment ids added to the visibleSegments set are tracked and updated as splits and merges are performed on the graph, including by other users.  There is a websocket connection to the server specifically for this purpose.

I would suppose graphene could support that as well, possibly with additional server-side support, but my understanding is that goes against the current model used in the seung-lab neuroglancer, where after selecting a segment, it is pinned to the version at the time it is selected (until a split/merge action is performed), and Sven indicated that this pinning behavior was considered desirable by users.

**Comment by chrisj** [comment]
Hi Jeremy,

Ran into an issue with implementing multicut using the temporary sets. This is an image of our multicut tool in action. As soon as I add a split point, the segment that contains the supervoxel becomes our focus segment. The supervoxel layer darkens everything outside of that focus segment. I have one supervoxel in the red set, one in the blue set. I am planning to use the stated colors to handle that coloring.

I tried to implement the "focus" feature in a similar way to our previous branch, using an additional hash table. The issue I run into seems to be that in order to filter by the focus segment, it needs to look up the non-temporary equivalence. I can set useTemporaryVisibleSegments to true and keep useTemporarySegmentEquivalences false. I can make something similar to valueForColor so it knows when to use value and base value for our scenario. The problem is this is starting to feel too specific to the situation for multicut. I was hoping everything I added would fairly generic.

Screen Shot 2021-09-22 at 4 35 07 PM

Basically I have a situation where I want to display some temporary and some non temporary.

**Comment by jbms** [comment]
Are you planning to still have a separate RenderLayer for the supervoxels?  I added the temporary segments options in order to avoid having to do that.  To be honest I don't remember exactly why I didn't want to use a separate RenderLayer for the temporary stuff shown during split/merge operations, but I recall that I initially attempted to go that route with nggraph but ran into some problems.

Do I understand correctly that the objective here is that once you start a multicut operation:

- Every segment except the focus segment should be hidden
- The focus segment should be shown in white
- The seed supervoxels should be shown in red and blue.

I think you could potentially accomplish this by:
- adding a temporary segment stated colors map, that works similarly to the existing temporary segments and temporary equivalences
- create a temporary equivalence map that links everything in the focus segment to its root id, except that it excludes the seed supervoxel ids
- add the root id and the seed supervoxel ids, to the temporary visible segments set
- add the root id and the seed supervoxel ids to the temporary segment color map

**Comment by chrisj** [comment]
I removed the separate render layer.
What you said is correct although it is more about darkening everything but the focus segment but perhaps that can be achieved with a combination of darken everything and brighten the focus segment.

I'll try out your suggestion, thanks!

**Comment by chrisj** [comment]
That seems like it will work well, I do think I need to add a render setting to darken all non-"visible segments".

I have another question for multicut. In the picture I posted earlier, you can also see a red annotation. We keep track of the position of the mouse clicks as I mentioned that we also do with the merge. For this tool we render them as an annotation layer. We did this previously by creating our own SliceViewAnnotationLayer and PerspectiveViewAnnotationLayer on the segmentation_user_layer for each the red and blue sets which means we don't display them as their own user layer. We display our own list of them separate in our graph tab separate from the layer annotations list.


I tried to see if I could do this with the `addAnnotationLayerState` or `addLocalAnnotations` on the SegmentationUserLayer which didn't exist when it was implemented in the seung lab branch. 


,

**Comment by jbms** [comment]
Yes, I think some refactoring of addLocalAnnotations/addAnnotationLayerState may be needed in order to allow these split/merge points to be shown as a separate annotation list.

As far as the units being wrong, moueState.position is always in the global coordinate space, but you should store these annotations in the coordinate space of the segmentation data source.  You can look at the `getMousePositionInAnnotationCoordinates` function defined in `ui/annotations.ts`.

**Comment by chrisj** [comment]
Does this seem like a reasonable way to create an AnnotationLayerState outside of activateDataSubsources?

My first solution was to add an optional method on graphConnection inside activateDataSubsources. It seemed messy but perhaps I could write a generic function to just allow graph connections to do whatever they need, passing the loadedSubsource and the layer.

Instead I decided to try a more brute force approach which I prefer because it allows me keep it from polluting segmentation_user_layer.

```typescript
function getGraphLoadedSubsource(layer: SegmentationUserLayer) {
  for (const dataSource of layer.dataSources) {
    const {loadState} = dataSource;
    if (loadState === undefined || loadState.error !== undefined) continue;
    for (const subsource of loadState.subsources) {
      if (subsource.enabled) {
        if (subsource.subsourceEntry.id === 'graph') {
          return subsource;
        }
      }
    }
  }
  return undefined;
}

function createAnnotationState(layer: SegmentationUserLayer) {  
  const loadedSubsource = getGraphLoadedSubsource(layer)!;
  const source = new LocalAnnotationSource(loadedSubsource.loadedDataSource.transform, [], []);
  const displayState = new AnnotationDisplayState();
  const state = new AnnotationLayerState({
      localPosition: layer.localPosition,
      transform: loadedSubsource.getRenderLayerTransform(),
      source,
      displayState,
      dataSource: loadedSubsource.loadedDataSource.layerDataSource,
      subsourceIndex: loadedSubsource.subsourceIndex,
      subsourceId: loadedSubsource.subsourceEntry.id,
      role: RenderLayerRole.ANNOTATION,
    });
  layer.addAnnotationLayerState(state, loadedSubsource);
  return state;
}
```
Inside my MulticutSegmentsTool constructor, I have
```typescript
    let initialized = false;
    this.layer.readyStateChanged.add(() => {
      if (!this.layer.isReady || initialized) return;
      initialized = true;
      createAnnotationState(this.layer);
      ...
```

**Comment by jbms** [comment]
I think your basic approach of adding the annotations is reasonable, given that they are really owned by the multicut tool, rather than by the datasource.

It would be good, though, to ensure they are only shown when the multicut tool is active, and are destroyed when the multicut tool is destroyed or the graph source is removed/deactivated.

Additionally, I think we don't want the points showing up in the normal annotation tab --- I suppose they should be shown in a separate list?

That would require modifying addAnnotationLayerState, so that they are somehow added to a different collection of annotation states, or something like that.  Maybe that can be deferred for now, though.

**Comment by chrisj** [comment]
@googlebot I consent

**Comment by kaikue** [comment]
@googlebot I consent.

**Comment by chrisj** [comment]
hey @jbms been a while since the last update and I've accumulated some more questions about design decisions. We've been making good progress though I still need to address many of your previous points. 

* First, one that I ran into a while ago and it looks like I never brought it up, is forEachVisibleSegment. We have meshes per root id, so with no changes, it ends up making a request for each equivalence and it slows down mesh loading. We just need it to run callback on each rootObjectId `callback(rootObjectId, rootObjectId)`.

    We don't set the high bit for our root ids, but if we did, the check for isHighBitSegment is doing the opposite of what we need for graphene. I'm not sure if it would make sense/possible to use the high bit for graphene. We have a graphene version of isBaseSegmentId and we could define a isRepresentitiveId though we rely on data from our MultiscaleVolumeInfo to calculate base/root segments, so it would need to be a property/method on GraphSource.

* I've implemented most of the Multicut functionality though I had to make some decisions I'm unsure about.

    The multicut state is stored in two AnnotationLayerStates/sources. I have a class that makes it more convenient to interact with them called MulticutState. Am I right that neuroglancer only saves annotations attached to annotation user layers? Not when I add them to the segmentaton user layer.

    Also, I attached this state to the SegmentationGraphSourceConnection so that it can be shared among several buttons attached to my GrapheneTab as well as the AnnotationLayerView on the GrapheneTab. The other buttons can possibly be removed, as I also added them to the makeToolActivationStatusMessageWithHeader but I still have the AnnotationLayerView. I see you added anchorSegment to SegmentationUserLayerDisplayState so maybe having the state attached to Tool isn't the right decision but I do feel like I just threw it on SegmentationGraphSourceConnection because I could.

* Few smaller things I wanted to mention. I extended AnnotationLayerView and overrode get annotationStates to control what states to display in our additional multicut annotation list. Feels a bit hacky but I only had to make the smallest change in AnnotationLayerView to support it.

    What is the reasoning behind this line? I wasn't sure about cases where the segmentationGroupState would change.

```ts
// Ensure we use the same segmentationGroupState while activated.
const segmentationGroupState = this.layer.displayState.segmentationGroupState.value;
```
    
I added the graphene tab using

```ts
if (segmentationGraph.tab) {
  const graphTab = segmentationGraph.tab;
  this.tabs.add(
    'graph', {label: 'Graph', order: -25, getter: () => {
      return graphTab(this);
    }});
  this.panels.updateTabs();
}
```

I think I should remove it in `refCounted.registerDisposer` though it looks like I would need to make some modifications to support removing tabs. Also, the tab adding is a bit crude so could use some input.

**Comment by jbms** [comment]
Hi Chris, thanks for your continued work on this!

Regarding `forEachVisibleSegment`: looks like maybe `highBitRepresentative` as a boolean value should be replaced by an enum that indicates one of several possible ways of iterating over the visible segments, e.g. `SIMPLE_EQUIVALENCES`, `HIGH_BIT_REPRESENTATIVE_EXCLUDED`, `HIGH_BIT_REPRESENTATIVE_ONLY`.

**Comment by jbms** [comment]
As far as storing the multicut seed points, you do need to explicitly store them somewhere in the json state if they are to be stored.  For annotation layers they are stored as an `annotations` property of the layer, but I think for this purpose it would make sense to store them within the json state of the multicut tool itself, since they are specific to that tool as far as I understand.  I made anchorSegment a property of the layer rather than the tool because it is used by both the split and merge tools, and it is convenient for them to share the same anchor rather than use separate ones.

**Comment by jbms** [comment]
Your tiny modification to AnnotationLayerView seems fine.

The segmentationGroupState is controlled by the "Linked segmentation" option on the "Rendering" tab.  That allows multiple segmentation layers to share the same set of selected segments and equivalences.  The goal there is to make sure we don't get into an inconsistent state if for some reason the user has changed it while the tool is activated.

As far as adding the additional Graph tab --- it is true that the current tab handling is not really designed for adding and removing tabs after the layer is initialized, and that will have to be fixed somehow.  A particular issue is what happens if the user has either selected the Graph tab, or dragged it out into a separate side panel.  In that case the tab won't initially exist when the layer is first created, because it is only added once the graphene data source is loaded.

**Comment by jbms** [comment]
Thinking a bit more about how to store the multicut points in the JSON state --- probably they should indeed be a property of the layer, storing it in the tool's own json state wouldn't work that well since that is associated with a specific key binding as well.

**Comment by chrisj** [comment]
Would it be possible to make graph connection trackable? Another idea I have is to make annotations on UserLayerWithAnnotationsMixin trackable, maybe an option when adding via addAnnotationLayerState. Though I would need to be able to identify the annotation and reconnect it with the tool. I think I would need to identify it. It probably wouldn't make sense for any of the annotations that are added from datasources.

Also, I forgot to mention my changes to segmentation_renderlayer so I could replicate what we did in the seung lab branch.

The following features only affect the 2d rendering.

* Added `baseSegmentHighlighting` (with UI) similar to `baseSegmentColoring` though it deals with the effect when hovering over a segment.

* Added `focusSegments` and `showFocusSegments` (no UI) so you can set them and toggle them. When a segment is focused, it has transparent color and non focus segments are colored black with 50%. I want segment stated colors to override the transparency of the focus segment because I then use segmentStatedColors to color supervoxels within the focus segment.

    A few things that would be clean up the code is a temporarySegmentStatedColors and adding transparency to segmentStated colors.

* Added `highlightColor` (with UI). Colors the current being hovered over rather than the brighten effect.  Unfortunately there is no way to clear it through the UI. Value of undefined disables this feature. It would also be nice if this supported transparency. Looks like it will take a while before input="color" supports that. The default opacity on works fine for our use. It may make sense to remove the UI, I could also make a color + checkbox widget to be able to toggle it.

If it would help, I could break these segmentation_renderlayer changes into a separate PR.

**Comment by jbms** [comment]
Change the `graphConnection` member of `SegmentationUserLayer` to be a `WatchableValue<SegmentationGraphSourceConnection|undefined>` seems reasonable.

Can you explain a bit more what you have in mind as far as making annotations trackable?

Also, perhaps you can give a high-level overview of how segments are displayed in both the 2d and 3d view when using the multicut tool?

As far as `focusSegments` --- if I understand correctly, in the seung-lab branch when you begin a multicut operation, the single agglomerated segment you are splitting is a "focus segment" and everything else is shown darker?  Is that the only purpose of the "focusSegments" functionality?  With the current local equivalences support, instead, everything but the single segment being split is hidden when the split tool is activated.

In general the tools should only manipulate "temporary" display properties, not the display properties that are persisted in the JSON state, as otherwise the JSON state can become corrupted if you copy it while using a tool, etc.   That was the reason for introducing the temporary properties.  If necessary, additional temporary display properties can be added.

As far as `baseSegmentHighlighting` --- I think it would be ideal if we can avoid having this exposed in the UI in addition to `baseSegmentColoring` unless it is really necessary.  It looks like this is enabled automatically by the multicut tool; that would be better handled by a "temporary" display property.

In general one of the goals of Neuroglancer is to try to have the 2-d and 3-d views correspond as much as possible.  It seems that the correspondence is lost as far as the colors and highlighting during graphene multicut operatiions; however, to support that would require encoding the base supervoxel ids in the meshes served by graphene so I can see why it is currently not done.

**Comment by chrisj** [comment]
When you add the first split point in multicut, the root segment is becomes the focus segment. (I made focus segment a set so it was more flexible but we also ever have one at a time). Removes all color, everything besides for the neuron is dark so the users know what they are splitting. Because we remove color, this frees us to color supervoxels red or blue if they are part of the sinks or sources.

We also use the highlight color so users know which group red/blue is currently active. We have an action to swap between the two groups and an action to add a point to the active group.

Here is an image of the multicut state in 2d, The clear sections show which root segment is in focus. I have 2 segments in the the sink/red group and 1 segment in the source/blue group.
Screen Shot 2021-11-08 at 5 30 34 PM

**Comment by jbms** [comment]
It seems like maybe there isn't a need for a separate set of `focusSegments`, and instead there could just be a temporary `notSelectedColor` option of some sort?

**Comment by jbms** [comment]
Also, it is true that at the moment it is kind of annoying to add additional temporary display options due to the large number of levels of watchable that they need to be threaded through.  Unfortunately I don't have a good solution for that at the moment.  Longer term I'm planning to see about replacing the watchable/signal mechanism in Neuroglancer with something based on MobX, which I think would significantly simplify the code and reduce the chance of forgetting to install appropriate signal callbacks, and hopefully not at any performance cost.  However, that would be an enormous refactor and not likely to happen soon.

**Comment by chrisj** [comment]
notSelectedColor for everything outside the focus segment? I first tried making the focus segment white though it definitely is harder to see the image data, which I think is one of the reasons we set the alpha to 0. I also think some people wouldn't enjoy the 3d being a white neuron but I can ask about that. 

As far as saving the multicut state, I was thinking graphConnection implement TrackableValue, not just watchable. Have segmentation_user_layer call this.graphConnection.toJSON and restoreState.

I don't know if I have anything insightful as far as how to save annotations. I just know the annotations added to segmentation_user_layer aren't saved on reload though annotations added via annotation_user_layer are saved. Maybe multicut could create annotation_user_layers. I'm not sure how multicut would reconnect with annotations on load.

**Comment by jbms** [comment]
Anything saved in the Neuroglancer state needs a specific location/path in the JSON representation.

For "local://annotations" annotation layers, there is an extra "annotations" property that encodes the annotations.

What do you have in mind as far as the graph connection state --- perhaps something like?

```json
"graphConnection": {
  "multicut": {
    "include": [
      {"position": [1, 2, 3], "segment": "123"}
    ],
    "exclude": [
      {"position": [4,5, 6], "segment": "456"}
    ]
  }
}
```

Alternatively "multicut" could just be a property of the layer itself.  Is there additional state you need to save in the graph connection?

**Comment by jbms** [comment]
You have a good point that white is not a good color choice for the cross-section view, and I guess maybe not ideal for the 3d view either.

An alternative would be to use a different color, like "green".

In general I am just keen to keep the correspondence between the 2d and 3d views as much as possible.

**Comment by chrisj** [comment]
something like that for the multicut state would work. We would also need to store the root id as well.

```json
"graphConnection": {
  "multicut": {
    "focusSegment": "789",
    "include": [
      {"position": [1, 2, 3], "segment": "123"}
    ],
    "exclude": [
      {"position": [4,5, 6], "segment": "456"}
    ]
  }
}
```

Besides multicut, we also want to store the time (if they are viewing a historical segmentation). Two points for our find path tool that I haven't added to this PR yet. Possibly the anchor segment for our merge and split, or we could add coordinates to go along with the existing anchorSegment property.

**Comment by chrisj** [comment]
Sven said that you had an idea on where we could store the multicut state. I believe he said something about the datasource metadata. If that's correct, how would that be done?

It was also mentioned a while aback about a way we could set default key bindings (and alter it so that shift wasn't necessary for some tool binds) with our build config. Is that functionality that we need to add or is it existing?

**Comment by jbms** [comment]
For the multicut state, I recommend storing it within the datasource, i.e.:

```json
{"layers": [
  {"name": "segmentation",
   "source": {
      "url": "graphene://...",
      "state": {"multicut": ...},
    },
  },
]
}
```

That will require changes to `layer_data_source.ts` to parse the state (e.g. `layerDataSourceSpecificationFromJson`) and probably various other places to pass the state around and handle the `toJSON` path as well.

Adding default key bindings in build config is functionality you will need to add.

**Comment by chrisj** [comment]
I'm back to focusing on this PR, I know we all want to wrap it up soon.

Does this look like the way you would implement states in datasources?
https://github.com/google/neuroglancer/pull/340/commits/93fb56b37bb2d0a39c0eb3237fb8b4bdd4a7592f

Also as far as default key bindings.
By default we could have it so that default key bindings get bound as long as a tool does not already have a key bound to it nor is the key bound to anything else.

In addition, we could have an option so that certain keybinds can consider the active layer, in which case they can be bound to multiple tools at the same time. We would bind those keys to all of the related tools as long as they don't have a key around bound to it. This feature seems like it would be more difficult to implement and you said it is a behavior you are trying to move away from.

I think for now we should focus on just the single tool binds. Our users are used to the selected layer affecting the keybinds though we aren't considering it critical.

You said "default key binding should occur when the layer is created". You said it would be tricky. Would it mean we have a path where a key can trigger a tool with no corresponding button (since it may not be created yet)?

**Comment by jbms** [comment]
If you just add some additional global "split" / "merge" key bindings (not using shift+letter) as an additional build-time option, to match the current behavior of the seung lab branch, then there should be no conflict.  I think that is the simplest approach since it involves the fewest changes.

The buttons themselves are just part of the UI panel, and are only created when the panel itself is displayed, so binding of keys should not be triggered by that.

The issue I was referring to with regard to layer creation is that when the layer is first created, the data sources have not yet been loaded.  Only once the data sources are loaded do we know whether there is a graphene source (at least under the current design, particularly since you can have a redirection so it can't be determined from the data source URL alone).  However, having key bindings change non-deterministically as data sources are loaded also seems unfortunate.

**Comment by chrisj** [comment]
I removed the graphene specific code from the core mesh classes. I added `getFragmentKey` to the MeshSource interface though we need it to return a fragment key and a fragment id. I'm not sure what to name it, I could split it into two separate functons `getFragmentKey` and `getFragmentId`. Graphene doesn't use the original fragment key because we want to re-use meshes even if the root id changes.

I also cleaned up the graph tab code though it still feels crudely implemented.

I also wanted to bring up my last iteration of saving the multicut state https://github.com/google/neuroglancer/pull/340/commits/03f9da11e4b8a7810169d19db03b99d548e5405c

**Comment by chrisj** [comment]
After talking to Sven, we decided to try to make. this PR more manageable, I'm going to break out the proofreading code to a separate PR.

**Comment by chrisj** [comment]
closing, is being split up into multiple PRs
